lib/logger/lib/logger/config.ex

```diff
@@ -139,69 +143,147 @@ defmodule Logger.Config do
     {:ok, state}
   end
 
-  ## Helpers
+  defp update_counter({counter, log, discard_threshold, discard_period}, periodic_check?) do
+    # If length is more than the total, it means the counter is behind,
+    # due to non-log messages, so we need to increase the counter.
+    #
+    # If length is less than the total, then we either have a spike or
+    # the counter drifted due to failures.
+    #
+    # Because we always bump the counter and then we send the message,
+    # there is a chance clients have bumped the counter but they did not
+    # deliver the message yet. Those bumps will be lost. At the same time,
+    # we are careful to read the counter first here, so if the counter is
+    # bumped after we read from it, those bumps won't be lost.
+    total = read_counter(counter)
+    {:message_queue_len, length} = Process.info(self(), :message_queue_len)
+    add_counter(counter, length - total)
+
+    # In case we are logging but we reached the threshold, we log that we
+    # started discarding messages. This can only be reverted by the periodic
+    # discard check.
+    cond do
+      total >= discard_threshold ->
+        if log == :log or periodic_check? do
+          warn("Attempted to log #{total} messages, which is above :discard_threshold")
+        end
+
+        {counter, :discard, discard_threshold, discard_period}
+
+      log == :discard and periodic_check? ->
+        warn("Attempted to log #{total} messages, which is below :discard_threshold")
+        {counter, :log, discard_threshold, discard_period}
+
+      true ->
+        {counter, log, discard_threshold, discard_period}
+    end
+  end
+
+  defp warn(message) do
+    utc_log = Application.fetch_env!(:logger, :utc_log)
+    event = {Logger, message, Logger.Utils.timestamp(utc_log), pid: self()}
+    :gen_event.notify(self(), {:warn, Process.group_leader(), event})
+  end
+
+  defp schedule_update_counter({_, _, _, discard_period}) do
+    Process.send_after(self(), {__MODULE__, :update_counter}, discard_period)
+  end
+
+  ## Counter Helpers
+
+  # TODO: Use counters exclusively when we require Erlang/OTP 22+.
+  defp new_counter() do
+    if Code.ensure_loaded?(:counters) do
+      {:counters, :counters.new(1, [:atomics])}
```

Comment:
Have you explored using the `write_concurrency` mode? As far as I understand the counters are not perfectly consistent anyway and provide more of a heuristic value rather than a hard limit.

Reply:
I did in the benchmarks. Because it is per row and we have a single row, write_concurrency actually make things worse.e

Reply:
We tested it here: https://github.com/elixir-lang/elixir/pull/8567#issuecomment-450852031

Reply:
@fishcakez there is no relevant  difference in the benchmarks linked above with the write_concurrency option compared to atomics.